### PR TITLE
refactor: track memory metrics in catalog

### PR DESCRIPTION
**Pull request by tustvold**
#1488 pushed the catalog memory tracking into the underlying chunk implementations.

The motivation for pushing the size tracking all the way down was an attempt to avoid it getting out of date, but not only did this not work in practice (#1311), but also as the catalog state machine has evolved the idea of chunks changing size without involving the catalog is unlikely.

Unfortunately the design has two rather confusing drawbacks:

* The memory usage reported by the catalog includes chunks that are in the process of being dropped and may not even exist in the catalog anymore
* The metrics for individual chunks contain a gauge that is passed into its constructor and corresponds to the catalogs metrics domain

Finally with #1959 drop is no longer a synchronous operation, and this creates challenges for the lifecycle policy where we don't want to drop more chunks than necessary.

This PR therefore reworks the catalog memory metrics to be tracked on the CatalogChunk, when a chunk is removed from the catalog its contribution is removed from the total immediately. With an async drop we can remove its contribution to the total at the point that the `Drop` lifecycle action is set (potentially restoring if the action fails).

If we at a later date want to know information about unreferenced / dropping chunks, we can add these as metrics on the underlying chunk implementations. We already have some examples of this in the read buffer.

### CODE REVIEW

File: mutable_buffer/src/chunk.rs [hunk 1]
```diff
@@ -48,9 +47,9 @@ pub enum Error {
 pub type Result<T, E = Error> = std::result::Result<T, E>;
 
 #[derive(Debug)]
+#[allow(missing_copy_implementations)]
 pub struct ChunkMetrics {
-    /// keep track of memory used by chunk
-    memory_bytes: GaugeValue,
+    // Placeholder
```
Comment (tustvold) [inline mutable_buffer/src/chunk.rs:52]: I opted to keep these as we will likely want to add metrics for the stores in the near-ish future, and it seemed silly to rip out the plumbing only to later put it back in...

File: server/src/db.rs [hunk 2]
```diff
@@ -1345,7 +1342,7 @@ mod tests {
             .eq(1.0)
             .unwrap();
 
-        catalog_chunk_size_bytes_metric_eq(&test_db.metric_registry, "mutable_buffer", 1143)
+        catalog_chunk_size_bytes_metric_eq(&test_db.metric_registry, "mutable_buffer", 1127)
```
Comment (tustvold) [inline server/src/db.rs:1345]: The GaugeValue is 16 bytes and was previously stored as part of the snapshot

File: server/src/db.rs [hunk 3]
```diff
@@ -1575,8 +1572,8 @@ mod tests {
             vec![] as Vec<u32>
         );
 
-        // verify size is reported until chunk dropped
-        catalog_chunk_size_bytes_metric_eq(&test_db.metric_registry, "read_buffer", 1486).unwrap();
+        // verify size is not accounted even though a reference to the chunk still exists
```
Comment (tustvold) [inline server/src/db.rs:1575]: This is the major behavior change

File: metrics/src/gauge.rs [hunk 4]
```diff
@@ -207,10 +207,10 @@ impl GaugeValue {
 
     /// Sets the local value for this GaugeValue
     pub fn set(&mut self, new: usize) {
-        if new > self.local {
-            self.inc(new - self.local)
-        } else {
-            self.decr(self.local - new)
+        match new.cmp(&self.local) {
```
Comment (tustvold) [inline metrics/src/gauge.rs:210]: This change isn't necessary for correctness but avoids some work when the value hasn't changed

File: server/src/db/catalog/chunk.rs [hunk 5]
```diff
@@ -325,15 +328,17 @@ impl CatalogChunk {
...
             addr,
             stage,
             lifecycle_action: None,
             metrics,
             time_of_first_write: None,
             time_of_last_write: None,
             time_closed: None,
-        }
+        };
+        chunk.record_write(); // The creation is considered the first and only "write"
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:340]: Why is this needed?

**Review by e-dard — APPROVED** [review]
LGTM 👍

File: server/src/db/catalog/chunk.rs [hunk 6]
```diff
@@ -385,6 +390,46 @@ impl CatalogChunk {
             self.time_of_first_write = Some(now);
         }
         self.time_of_last_write = Some(now);
+        self.update_memory_metrics()
+    }
+
+    fn update_memory_metrics(&mut self) {
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:396]: Can we please add a docstring here explaining what this does (and that it should be called every time `self.stage` is changed?). 

I do worry about bugs being introduced in the future by "forgetting to call `update_memory_metrics` when `stage` is changed.

One way to potentially try and make that sort of thing harder to write would be to restructure the code so `self.stage = ...` was replaced with a call to `self.set_stage(...)` type function that would also include the call to `update_memory_metrics`

File: mutable_buffer/src/chunk.rs [hunk 7]
```diff
@@ -48,9 +47,9 @@ pub enum Error {
 pub type Result<T, E = Error> = std::result::Result<T, E>;
 
 #[derive(Debug)]
+#[allow(missing_copy_implementations)]
```
[the remark this answers is not in the document]
Comment (crepererum) [reply mutable_buffer/src/chunk.rs:50]: Yes: the copy lint suggest to impl Copy when the type is simple enough, which currently is the case but won't be in the future. I had the same issue in some other places and an explicit opt-out is a good solution to that issue.

File: server/src/db/catalog/chunk.rs [hunk 8]
```diff
@@ -712,6 +760,7 @@ impl CatalogChunk {
                             parquet: chunk,
                             read_buffer: Some(db),
                         };
+                        self.update_memory_metrics();
```
Comment (crepererum) [inline server/src/db/catalog/chunk.rs:763]: these explicit calls are something that I can see will be forgotten at some point, but I also cannot think of any good alternative ... :thinking: we might just enforce the update after every `&mut self` usage, but that also seems a bit overkill from a type-state / macro POV

**Review by alamb — APPROVED** [review]
FWIW the metrics handling looks much nicer now . Nice work

> when a chunk is removed from the catalog its contribution is removed from the total immediately.

Where are the metrics decremented when a chunk is dropped? I didn't see any explicit call. Is it implicit as part of `ChunkMetrics` when that is dropped? 

I double checked and the lifecycle policy uses [`buffer_size()`](https://github.com/influxdata/influxdb_iox/blob/track-memory-metrics-in-catalog/server/src/db/lifecycle.rs#L240) to calculate the total memory used which already was using the catalog metrics. 👍 

> With an async drop we can remove its contribution to the total at the point that the Drop lifecycle action is set (potentially restoring if the action fails).

Just to confirm I understand, the idea is that @crepererum  will update https://github.com/influxdata/influxdb_iox/pull/1959 (that introduces the `Drop` lifecycle) so that the accounting is updated when the `Drop` lifecycle action is set

File: server/src/db.rs (hunk 3 above)
Comment (crepererum) [reply server/src/db.rs:1575]: how does the metric know that the chunk was removed from the catalog?

**Review by crepererum — COMMENTED** [review]
I'm not sure that solves the issue w/ the async drop job, because what it does is:

- pre-checks
- perform persisted drop, includes IO, might fail
- if IO worked, perform the in-mem catalog drop

Is the plan the we call "update metrics" when the drop lifecycle action starts?

File: server/src/db.rs (hunk 3 above)
Comment (tustvold) [reply server/src/db.rs:1575]: When you drop a GaugeValue it decrements any contribution it made to the total

**Comment by tustvold** [comment]
> Is the plan the we call "update metrics" when the drop lifecycle action starts?

Yes, when you set the lifecycle action on the chunk you zero out the Gauge. If the lifecycle action is later cleared, e.g. because it failed, you could restore the metric count in `clear_lifecycle_action`. The totals effectively don't then include dropping chunks

File: server/src/db/catalog/chunk.rs (hunk 5 above)
Comment (tustvold) [reply server/src/db/catalog/chunk.rs:340]: It needs to at least call `update_memory_metrics` and the current formulation of the write timestamps is that they are the time of writes to the catalog chunk, not the rows themselves. This is therefore consistent with the current code for things like the read buffer chunk, etc...

Comment (alamb) [reply server/src/db/catalog/chunk.rs:340]: I just found it quite confusing as this wasn't a write. Calling `update_memory_metrics` makes sense to me

File: server/src/db/catalog/chunk.rs (hunk 8 above)
Comment (alamb) [reply server/src/db/catalog/chunk.rs:763]: I suggest something like
```rust
  fn set_state(&mut self, new_state: State) {
    self.state = new_state;
    self.update_memory_metrics();
  }
```

Not the most high tech / programming language fancy way, but I suspect it might catch most mis-uses

File: server/src/db.rs (hunk 3 above)
Comment (crepererum) [reply server/src/db.rs:1575]: So `reference to the chunk` means `reference to the DbChunk` here, while the `CatalogChunk` was dropped?

Comment (tustvold) [reply server/src/db.rs:1575]: Oh yeah, I see the confusion, specifically a reference to the RubChunk exists via the DbChunk - I will update the comment

**Review by crepererum — APPROVED** [review]

File: server/src/db/catalog/chunk.rs (hunk 8 above)
Comment (tustvold) [reply server/src/db/catalog/chunk.rs:767]: So a lot of the code doesn't directly manipulate the state, it instead modified members of the state, e.g. representation. 

I tried doing this, but it just creates an even weirder situation where set_state updates some of the metrics, with some left for the state transition methods. I'm going to leave this as is and potentially revisit if I think of a better way to do this
